res/res_stir_shaken/verification.c

```diff
@@ -0,0 +1,1033 @@
+			ctx->tag, ctx->filename, ctx->public_url);
+	}
+
+	vs_rc = check_cert(ctx);
+	if (vs_rc != AST_STIR_SHAKEN_VS_SUCCESS) {
+		X509_free(ctx->xcert);
+		ctx->xcert = NULL;
+		SCOPE_EXIT_RTN_VALUE(vs_rc, "%s: Cert '%s' failed validity checks\n",
+			ctx->tag, ctx->public_url);
+	}
+
+	SCOPE_EXIT_RTN_VALUE(AST_STIR_SHAKEN_VS_SUCCESS,
+		"%s: Cert '%s' successfully retrieved from cache\n",
+		ctx->tag, ctx->public_url);
+}
+
+static enum ast_stir_shaken_vs_response_code ctx_populate(
+	struct ast_stir_shaken_vs_ctx *ctx)
+{
+	char hash[41];
+
+	ast_sha1_hash(hash, ctx->public_url);
+	if (ast_string_field_set(ctx, hash, hash) != 0) {
+		return AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+	}
+
+	if (ast_string_field_build(ctx, filename, "%s/%s.pem",
+		ctx->vs->cert_cache_dir, hash) != 0) {
+		return AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+	}
+
+	if (ast_string_field_build(ctx, hash_family, "%s/hash/%s",
+		AST_DB_FAMILY, hash) != 0) {
+		return AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+	}
+
+	if (ast_string_field_build(ctx, url_family, "%s/url", AST_DB_FAMILY) != 0) {
+		return AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+	}
+
+	return AST_STIR_SHAKEN_VS_SUCCESS;
+}
+
+static enum ast_stir_shaken_vs_response_code
+	retrieve_verification_cert(struct ast_stir_shaken_vs_ctx *ctx)
+{
+	enum ast_stir_shaken_vs_response_code rc = AST_STIR_SHAKEN_VS_SUCCESS;
+	SCOPE_ENTER(3, "%s: Retrieving cert '%s'\n", ctx->tag, ctx->public_url);
+
+	ast_trace(1, "%s: Checking cache for cert '%s'\n", ctx->tag, ctx->public_url);
+	rc = retrieve_cert_from_cache(ctx);
+	if (rc == AST_STIR_SHAKEN_VS_SUCCESS) {
+		SCOPE_EXIT_RTN_VALUE(rc, "%s: Using cert '%s' from cache\n",
+			ctx->tag, ctx->public_url);;
+	}
+
+	ast_trace(1, "%s: No valid cert for '%s' available in cache\n",
+		ctx->tag, ctx->public_url);
+	ast_trace(1, "%s: Retrieving cert directly from url '%s'\n",
+		ctx->tag, ctx->public_url);
+
+	rc = retrieve_cert_from_url(ctx);
+	if (rc == AST_STIR_SHAKEN_VS_SUCCESS) {
+		SCOPE_EXIT_RTN_VALUE(rc, "%s: Using cert '%s' from internet\n",
+			ctx->tag, ctx->public_url);
+	}
+
+	SCOPE_EXIT_LOG_RTN_VALUE(rc, LOG_ERROR,
+		"%s: Unable to retrieve cert '%s' from cache or internet\n",
+		ctx->tag, ctx->public_url);
+}
+
+enum ast_stir_shaken_vs_response_code
+	ast_stir_shaken_vs_ctx_add_identity_hdr(
+	struct ast_stir_shaken_vs_ctx * ctx, const char *identity_hdr)
+{
+	return ast_string_field_set(ctx, identity_hdr, identity_hdr) == 0 ?
+		AST_STIR_SHAKEN_VS_SUCCESS : AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+}
+
+enum ast_stir_shaken_vs_response_code
+	ast_stir_shaken_vs_ctx_add_date_hdr(struct ast_stir_shaken_vs_ctx * ctx,
+	const char *date_hdr)
+{
+	return ast_string_field_set(ctx, date_hdr, date_hdr) == 0 ?
+		AST_STIR_SHAKEN_VS_SUCCESS : AST_STIR_SHAKEN_VS_INTERNAL_ERROR;
+}
+
+enum ast_stir_shaken_failure_action
+	ast_stir_shaken_vs_get_failure_action(
+		struct ast_stir_shaken_vs_ctx *ctx)
+{
+	return ctx->failure_action;
+}
+
+int	ast_stir_shaken_vs_get_use_rfc9410_responses(
```

Comment:
Looks like a `\t` here after `int` instead of a space

Reply:
done